### PR TITLE
Convert more code to TypeScript

### PR DESCRIPTION
**Pull request by martikaljuve**
Hello.

I took the `master` branch, converted JS files to TS and added more strong types. This should make further development easier.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,14 +27,15 @@
...
+    "@types/pako": "2.0.3",
 
     "rollup": "2.58.0",
     "rollup-plugin-babel": "4.4.0",
     "rollup-plugin-commonjs": "10.1.0",
     "rollup-plugin-node-resolve": "5.2.0",
     "rollup-plugin-terser": "7.0.2",
     "rollup-plugin-typescript2": "0.33.0",
-    "typescript": "4.8.2",
+    "typescript": "4.9.5",
```
Comment (martikaljuve) [inline package.json:38]: Updated TypeScript to `4.9.x` so I could use the [satisfies](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator) operator. Could also consider updating to latest `5.x`, but that would probably require updating ESLint libraries as well.

File: src/lib/RawMapData.ts [hunk 2]
```diff
@@ -45,23 +47,18 @@ export interface RawMapLayerMetaData {
...
-    GoToTarget = "go_to_target",
-    Path = "path",
-    PredictedPath = "predicted_path",
-    VirtualWall = "virtual_wall",
-    NoGoArea = "no_go_area",
-    NoMopArea = "no_mop_area",
-    ActiveZone = "active_zone",
-}
+export type RawMapEntityType =
+    | "charger_location"
```
Comment (martikaljuve) [inline src/lib/RawMapData.ts:53]: Replaced enum with a simple union, as the rest of the code doesn't use enums anyway (e.g. `RawMapEntityType.ChargerLocation`).

File: src/valetudo-map-card.ts [hunk 3]
```diff
@@ -1,9 +1,14 @@
-import {FourColorTheoremSolver} from "./ts-lib/colors/FourColorTheoremSolver";
 import * as pako from "pako";
-import {extractZtxtPngChunks} from "./lib/pngUtils";
-import {DEFAULT_CARD_CONFIG, POLL_INTERVAL_STATE_MAP} from "./res/consts";
-import {preprocessMap} from "./ts-lib/mapUtils";
+import { HomeAssistant } from "custom-card-helpers";
```
Comment (Hypfer) [inline src/valetudo-map-card.ts:2]: Are these missing in the package.json?

File: src/res/consts.ts [hunk 4]
```diff
@@ -64,9 +66,9 @@ export const DEFAULT_CARD_CONFIG = Object.freeze({
 
     // Crop settings
     min_height: 0
-});
+} satisfies Partial<Configuration>);
 
-export const POLL_INTERVAL_STATE_MAP = Object.freeze({
+export const POLL_INTERVAL_STATE_MAP: Record<string, number> = Object.freeze({
```
Comment (martikaljuve) [inline src/res/consts.ts:71]: Could also use a more advanced type here:

```ts
type RobotState = "cleaning" | "docked" | "error" | "idle" | "paused" | "returning" | "unavailable" | "unknown";

export const POLL_INTERVAL_STATE_MAP: Partial<Record<RobotState, number>> = Object.freeze({
  /../
```

File: src/valetudo-map-card.ts (hunk 3 above)
Comment (martikaljuve) [reply src/valetudo-map-card.ts:2]: Yes, `home-assistant-js-websocket` is a subdependency of `custom-card-helpers`, will add that explicitly to package.json as well. 👍

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
+    "compilerOptions": {
```
Comment (martikaljuve) [inline tsconfig.json:2]: Could consider adding `"strict": true` to the tsconfig as well.

**Pull request merged**